services/src/main/java/org/keycloak/services/resources/admin/GroupsResource.java

```diff
@@ -94,21 +94,25 @@ public Stream<GroupRepresentation> getGroups(@QueryParam("search") String search
         GroupPermissionEvaluator groupsEvaluator = auth.groups();
         groupsEvaluator.requireList();
 
-        Stream<GroupModel> stream = null;
+        Stream<GroupModel> stream;
         if (Objects.nonNull(searchQuery)) {
             Map<String, String> attributes = SearchQueryUtils.getFields(searchQuery);
-            stream = ModelToRepresentation.searchGroupModelsByAttributes(session, realm, !briefRepresentation, populateHierarchy, attributes, firstResult, maxResults);
+            stream = ModelToRepresentation.searchGroupModelsByAttributes(session, realm, attributes, firstResult, maxResults);
         } else if (Objects.nonNull(search)) {
-            stream = ModelToRepresentation.searchForGroupModelByName(session, realm, !briefRepresentation, search.trim(), exact, firstResult, maxResults);
+            stream = session.groups().searchForGroupByNameStream(realm, search.trim(), exact, firstResult, maxResults);
         } else if(Objects.nonNull(firstResult) && Objects.nonNull(maxResults)) {
-            stream = ModelToRepresentation.toGroupModelHierarchy(realm, !briefRepresentation, firstResult, maxResults);
+            stream = realm.getTopLevelGroupsStream(firstResult, maxResults);
         } else {
             stream = realm.getTopLevelGroupsStream();
         }
 
-        boolean canViewGlobal = groupsEvaluator.canView();
-        return stream.filter(group -> canViewGlobal || groupsEvaluator.canView(group))
-                .map(group -> GroupUtils.toGroupHierarchy(groupsEvaluator, group, search, exact, !briefRepresentation, false));
+        if(populateHierarchy) {
+            return GroupUtils.populateGroupHierarchyFromSubGroups(session, realm, stream, !briefRepresentation, groupsEvaluator);
+        }
+        return stream
+            .filter(g -> groupsEvaluator.canView() || groupsEvaluator.canView(g))
```

Comment:
Maybe it is minor comment, but why the variable `canViewGlobal` was removed and there is introduced call to `groupsEvaluator.canView()` for every group? The result would be always the same for all groups, so looks like unnecessary performance overhead (at least in theory).

Reply:
totally agree! Oversight on my part I think

Reply:
Thanks! Resolving this comment